### PR TITLE
Limiting query start time with config

### PR DESCRIPTION
**Pull request by sandlis**
Added maxLookBackPeriod in config to limit query start time

### CODE REVIEW

**Comment by daixiang0** [comment]
please add this option into all config files.

**Comment by sandlis** [comment]
> please add this option into all config files.

Sorry didn't get you. Do you mean add it to yaml config in cmd/loki?

**Comment by daixiang0** [comment]
yep.

File: pkg/querier/querier.go [hunk 1]
```diff
@@ -18,10 +19,12 @@ import (
 
 // Config for a querier.
 type Config struct {
+	MaxLookBackPeriod time.Duration `yaml:"max_look_back_period"`
 }
 
 // RegisterFlags register flags.
 func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	f.DurationVar(&cfg.MaxLookBackPeriod, "querier.max-look_back_period", 0, "")
```
Comment (gouthamve) [inline pkg/querier/querier.go]: Why the mixure of `-` and `_`? Also, can you add a description for the flag?

File: pkg/querier/querier.go [hunk 2]
```diff
@@ -18,10 +19,12 @@ import (
 
 // Config for a querier.
 type Config struct {
+	MaxLookBackPeriod time.Duration `yaml:"max_look_back_period"`
```
Comment (gouthamve) [inline pkg/querier/querier.go]: Can you add a comment above the field too?

File: pkg/querier/querier.go [hunk 3]
```diff
@@ -92,6 +95,11 @@ func (q *Querier) forAllIngesters(f func(logproto.QuerierClient) (interface{}, e
 
 // Query does the heavy lifting for an actual query.
 func (q *Querier) Query(ctx context.Context, req *logproto.QueryRequest) (*logproto.QueryResponse, error) {
+	oldestStartTime := time.Now().Add(-q.cfg.MaxLookBackPeriod)
```
Comment (gouthamve) [inline pkg/querier/querier.go]: This means queries are broken for the default value (0). Have a check for `MaxLookBackPeriod > 0` :)

File: pkg/querier/querier.go (hunk 1 above)
Comment (sandlis) [reply pkg/querier/querier.go]: must be a typo, sorry missed it somehow

**Comment by sandlis** [comment]
> Right now we cannot use the config file to configure this.

@gouthamve Didn't get what you meant with `Right now we cannot use the config file to configure this.`

**Comment by gouthamve** [comment]
Right, nvm, I missed it, I was thinking `querier.Config` is not registered to be in the YAML. It is. I think what @daixiang0 meant seems to be about updating the helm chart to have this option available.

File: production/helm/loki/values.yaml [hunk 4]
```diff
@@ -56,6 +56,8 @@ config:
       directory: /data/loki/index
     filesystem:
       directory: /data/loki/chunks
+  querier:
+    max_look_back_period: '720h'
```
Comment (daixiang0) [inline production/helm/loki/values.yaml]: why different default value?

Comment (daixiang0) [reply production/helm/loki/values.yaml]: And no need to quote it.

Comment (sandlis) [reply production/helm/loki/values.yaml]: You mean this should be 0? Same as what is set in RegisterFlags [here](https://github.com/sandlis/loki/blob/query-max-lookback/pkg/querier/querier.go#L28)

Comment (gouthamve) [reply production/helm/loki/values.yaml]: Yes, the defaults should be the same, one of the reasons we are adding it to helm is to make it configurable by the user, but if the user doesn't change anything, loki should start with the defaults.

File: cmd/loki/loki-local-config.yaml [hunk 5]
```diff
@@ -30,3 +30,6 @@ storage_config:
 
 limits_config:
   enforce_metric_name: false
+
+querier:
+  max_look_back_period: 672h
```
Comment (gouthamve) [inline cmd/loki/loki-local-config.yaml]: Can you change this also to `0`?

File: pkg/querier/querier.go [hunk 6]
```diff
@@ -18,10 +19,13 @@ import (
 
 // Config for a querier.
 type Config struct {
+	// Limits query start time to be greater than now() - MaxLookBackPeriod, if set
```
Comment (gouthamve) [inline pkg/querier/querier.go]: Comments should end with `.`

**Pull request merged**